src/storage/src/client/controller.rs

```diff
@@ -485,12 +498,27 @@ where
                 .insert_without_overwrite(&mut self.state.stash, &id, metadata)
                 .await?;
 
-            let (write, read) = self
+            let (mut write, mut read) = self
                 .persist_client
                 .open(metadata.data_shard)
                 .await
                 .expect("invalid persist usage");
 
+            // Advance the collection's `since` and `upper` to the given `as_of`.
+            if let Some(as_of) = &description.as_of {
+                let lower = write.upper().clone();
+                if PartialOrder::less_than(&lower, as_of) {
```

Comment:
You could remove the `if` because the `append` is a no-op if the upper is already beyond the `as_of`. Though that would do at least one persist call more. So it's fine to keep it like this.

Reply:
Ah, I thought this check would make `append` return an error if lower > upper: https://github.com/MaterializeInc/materialize/blob/fb6244a96848b36e49a39d09fa167b128e8426ec/src/persist-client/src/batch.rs#L236

I haven't tested it, admittedly, will do that tomorrow. I've we don't need that `if`, I'm all for removing it!